src/ui/filter/security.go

```diff
@@ -40,20 +48,88 @@ func SecurityFilter(ctx *context.Context) {
 		return
 	}
 
-	if !strings.HasPrefix(req.RequestURI, "/api/") &&
-		!strings.HasPrefix(req.RequestURI, "/service/") {
+	if !strings.HasPrefix(req.URL.RequestURI(), "/api/") &&
+		!strings.HasPrefix(req.URL.RequestURI(), "/service/") {
 		return
 	}
 
-	securityCtx, err := createSecurityContext(req)
-	if err != nil {
-		log.Warningf("failed to create security context: %v", err)
+	// fill ctx with security context and project manager
+	fillContext(ctx)
+}
+
+func fillContext(ctx *context.Context) {
+	// secret
+	scrt := ctx.GetCookie("secret")
+	if len(scrt) != 0 {
+		ctx.Input.SetData(HarborProjectManager,
+			getProjectManager())
+
+		log.Info("creating a secret security context...")
+		ctx.Input.SetData(HarborSecurityContext,
+			secret.NewSecurityContext(scrt, config.SecretStore))
+
 		return
 	}
 
-	ctx.Input.SetData(HarborSecurityContext, securityCtx)
+	var user *models.User
+	var err error
+
+	// session
+	username := ctx.Input.Session("username")
+	if username != nil {
+		user, err = dao.GetUser(models.User{
+			Username: username.(string),
+		})
+		if err != nil {
+			log.Errorf("failed to get user %s: %v", username.(string), err)
+		}
+
+		if user != nil {
+			log.Info("get credential from session")
+		}
+	}
+
+	// basic auth
+	if user == nil {
+		username, password, ok := ctx.Request.BasicAuth()
+		if ok {
+			// TODO the return data contains other params when integrated
+			// with vic
+			user, err = auth.Login(models.AuthModel{
+				Principal: username,
+				Password:  password,
+			})
+			if err != nil {
+				log.Errorf("failed to authenticate %s: %v", username, err)
+			}
+			if user != nil {
+				log.Info("get credential from basic auth")
+			}
+		}
+	}
+
+	if user == nil {
+		log.Info("get no credential")
+	}
+
+	pm := getProjectManager()
+	ctx.Input.SetData(HarborProjectManager, pm)
+
+	log.Info("creating a rbac security context...")
+	ctx.Input.SetData(HarborSecurityContext,
+		rbac.NewSecurityContext(user, pm))
+
+	return
 }
 
-func createSecurityContext(req *http.Request) (security.Context, error) {
-	return nil, nil
+func getProjectManager(token ...string) projectmanager.ProjectManager {
```

Comment:
I think you should use request as input parm, just in case we put token in session?